backup/queries_statistics.go

```diff
@@ -119,13 +120,20 @@ func GetAttributeStatistics(connectionPool *dbconn.DBConn, tables []Table) map[u
 		inheritClause, statSlotClause, statCollationClause,
 		SchemaFilterClause("n"), utils.SliceToQuotedString(tablenames))
 
-	results := make([]AttributeStatistic, 0)
-	err := connectionPool.Select(&results, query)
+	return connectionPool.Query(query)
+}
+
+func GetAttributeStatistics(connectionPool *dbconn.DBConn, tables []Table) map[uint32][]AttributeStatistic {
+	results, err := GetAttributeStatisticsRows(connectionPool, tables)
 	gplog.FatalOnError(err)
 	stats := make(map[uint32][]AttributeStatistic)
-	for _, stat := range results {
-		stats[stat.Oid] = append(stats[stat.Oid], stat)
+	for results.Next() {
```

Comment:
How fetching is implemented? Because `Query` in fact is [done through](https://github.com/jackc/pgconn/blob/master/pgconn.go#L826) PREPARE STATEMENT. Is fetching using that sheme is equivalent to the fetching from declared cursor?

Reply:
> Can we be confident in lower memory consumption?

my testing scenario shows ~ 150 MB instead ~ 3 GB

Reply:
> Is fetching using that sheme is equivalent to the fetching from declared cursor?

No, but much memory is used when filling resulting map at once

Reply:
Does result of query is materialized somewhere on the server side? Or the mechanism is more complex?

Reply:
>my testing scenario shows ~ 150 MB instead ~ 3 GB

150 MB represents size of what? memory connsumption by gpbackup during statistics backup?

Reply:
> 150 MB represents size of what? memory connsumption by gpbackup during statistics backup?

yes, I see in htop

Reply:
> Is result of query materialized somewhere on the server side? Or the mechanism is more complex?

before patch we load all rows in map, that has ~ 3GB size on my test, after patch we load only one row into struct and save it to sql file, that leads to low memory consumption

Reply:
The sqlx.Queryx function for [our case](https://github.com/arenadata/gpbackup/blob/d89c01996449c77576a0f1e2f8a169bbe0bec853/backup/queries_statistics.go#L166) is calling [sql.ctxDriverQuery](https://github.com/golang/go/blob/62711d61e5ba3a6c42bc57a28c2402dd8669261f/src/database/sql/ctxutil.go#L60), which calls the Query method from pq driver. pq driver eventualy [works with prepare/execute](https://github.com/lib/pq/blob/3d613208bca2e74f2a20e04126ed30bcb5c4cc27/conn.go#L917-L918) scheme. Execute step processes [all rows](https://github.com/arenadata/gpdb/blob/7ef8cd7199151c8a64d48f97825340763ca43c58/src/backend/tcop/pquery.c#L1162-L1173) (count = 0) of the input query. After each slot is extracted, its data is [written](https://github.com/arenadata/gpdb/blob/7ef8cd7199151c8a64d48f97825340763ca43c58/src/backend/libpq/pqcomm.c#L1482-L1494) into PqSendBuffer. When the size of written data exceeds 8K, the buffer is flushed, the data is written to socket. The Rowsx.Next() probably fetches rows staight from conenction.

Thats results of my investiagtion how data is processed. Any thoughts? Do you agree that this technique is suitable? Will such scheme have any pitfalls?

Reply:
> Do you agree that this technique is suitable?

yes, because the main problem is the large map in Go for storing all the rows at once